### DISCUSSION TITLE
Fork for microntrollers

### DISCUSSION THREAD

**Discussion by rprata** [edubart/nelua-lang#42] [Q&A] [answered]
Hi maintainers, 

I've forked this project and I'm implementing support to develop applications for `Arduino Uno` using `nelua`. Is it good for the main project (I could create a pull request with my modifications) or is better to keep this fork separated of main repository (with a different name, but updated with upstream)?

**Comment by edubart** [comment, marked as answer]
I don't think a fork is need to improve support for Arduino, probably you want just to make a separate library and request features or potential fixes to improve the Nelua compiler usage with Arduino.
But you are free to fork and do whatever you want as the project is open source.

Regarding the pull requests, Nelua is not a language for any specific platforms, just like C/Lua isn't, so I probably wouldn't merge specific APIs/modules/tools for Arduino, as the language try to remain platform agnostic. But I would merge fixes and improved flexibility to make coding on platforms such as Arduino easier. In any case it would depend very much on the pull request.

**Comment by rprata** [comment]
@edubart , 
Thank you for your answer. I've tried to create a separate library for Arduino, but I didn't success. I needed to adjust some scripts:
- `nelua/cgenerator.lua`
- `nelua/runner.lua`

This changes were applied because Arduino Uno architecture (8 bits) is not compatible with nelua.

So, is it possible to send my modifications for this context, not harming the project completely (creating a conditional for this context)?

**Comment by rprata** [comment]
@edubart , thank you. I adjusting according your suggestions and I could revert my modifications. In this case I'll create a separeted library for arduino development.

**Comment by technomancy** [comment]
I'm interested in using nelua for microcontrollers too, but maybe without Arduino.

I see plenty of examples of how to call C functions, but are there any examples of setting memory locations directly? I have some code that directly manipulates ports like `PORTB = 0xff` and I'm wondering how to do that from nelua.

**Reply by edubart** [reply]
> I'm interested in using nelua for microcontrollers too, but maybe without Arduino.
> 
> I see plenty of examples of how to call C functions, but are there any examples of setting memory locations directly? I have some code that directly manipulates ports like `PORTB = 0xff` and I'm wondering how to do that from nelua.

In that case you have to import the symbol with the correct type, so the compiler knows, I am assuming you are talking about AVR, here is the following example:

```lua
local PORTB: uint8 <cimport'PORTB',nodecl,cinclude'<avr/io.h>'>

local function main(): cint <entrypoint>
  PORTB = 0b11111111 -- should be equivalent to 0xff
  return 0
end
```

The above code should generate this C code:
```C
/* ------------------------------ DECLARATIONS ------------------------------ */
#include <avr/io.h>
#include <stdint.h>
int main();
/* ------------------------------ DEFINITIONS ------------------------------- */
int main() {
  PORTB = (uint8_t)0xffU;
  return 0;
}
```

When compiling with this line:
```C
$ nelua --cc=avr-gcc --cpu-bits=16 --cflags="-mmcu=atmega2560" --verbose --compile-binary test-avr.nelua
generated /home/bart/projects/nelua/nelua-lang/nelua_cache/test-avr.c
avr-gcc "/home/bart/projects/nelua/nelua-lang/nelua_cache/test-avr.c" -o "/home/bart/projects/nelua/nelua-lang/nelua_cache/test-avr" -Wall -fwrapv -g -mmcu=atmega2560
```

Note the `--cpu-bits=16` line, this is important so the compiler assumes the correct sizes for the primitive types (e.g size of a pointer is 2), otherwise your code won't compile.

**Reply by technomancy** [reply]
Brilliant; thanks for the comprehensive answer. Will give this a try.

Is this something that should be covered in the "C Interoperability" section of the docs or is there somewhere more appropriate? https://nelua.io/overview/#c-interoperability

**Reply by technomancy** [reply]
I'm running into some trouble when I try to put a reference in an array here:

```lua
local PORTB: uint8 <cimport'PORTB',nodecl,cinclude'<avr/io.h>'>
local PORTC: uint8 <cimport'PORTB',nodecl,cinclude'<avr/io.h>'>
local PORTD: uint8 <cimport'PORTB',nodecl,cinclude'<avr/io.h>'>
local PORTE: uint8 <cimport'PORTB',nodecl,cinclude'<avr/io.h>'>
local PORTF: uint8 <cimport'PORTB',nodecl,cinclude'<avr/io.h>'>

local col_ports: array(pointer(uint8), 11) =
   {&PORTB, &PORTD, &PORTF, &PORTF, &PORTB,
       &PORTD,
    &PORTE, &PORTB, &PORTB, &PORTC, &PORTD}

-- $ nelua --cpu-bits=16 --cflags="-mmcu=atmega32u4 -D F_CPU=16000000UL -Os usb_keyboard.o" --cc=avr-gcc --verbose --compile-binary test.nelua
-- generated nelua_cache/test.c
-- avr-gcc "nelua_cache/test.c" -o "nelua_cache/test" -Wall -fwrapv -g -mmcu=atmega32u4 -D F_CPU=16000000UL -Os usb_keyboard.o
-- C compilation for 'nelua_cache/test' failed:
-- nelua_cache/test.c:23:29: error: static assertion failed: "Nelua and C disagree on type size or align"
--  #define nelua_static_assert _Static_assert
--                              ^
-- nelua_cache/test.c:31:59: note: in expansion of macro ‘nelua_static_assert’
--  typedef struct {nluint8_ptr data[11];} nluint8_ptr_arr11; nelua_static_assert(sizeof(nluint8_ptr_arr11) == 22 && _Alignof(nluint8_ptr_arr11) == 2, "Nelua and C disagree on type size or align");
```

Am I doing something wrong?

**Comment by edubart** [comment]
@technomancy At the time you tried the above Nelua support for AVR was not in good shape and since you asked I've made many improvements to make the compiler to work better with AVR. Now AVR passed all possible test from the test suite, the --cpu-bits=16 flag is not needed anymore and you should not get errors with "Nelua and C disagree on type size or align" anymore. Also many parts of Nelua standard library are more freestanding, that is, they use very few functions from lib C now, because AVR does not support some lib C functions.
